Shortcut query phase using the results of other shards

jimczi opened the pull request:
This commit, built on top of #51708, allows to modify shard search requests based on informations collected on other shards. It is intended to speed up sorted queries on time-based indices. For queries that are only interested in the top documents.
 
This change will rewrite the shard queries to match none if the bottom sort value computed in prior shards is better than all values in the shard. 
For queries that mix top documents and aggregations this change will reset the size of the top documents to 0 instead of rewriting to match none. 
This means that we don't need to keep a search context open for this shard since we know in advance that it doesn't contain any competitive hit.

Closes #49601

jimczi committed:
Always rewrite search shard request outside of the search thread pool

This change ensures that the rewrite of the shard request is executed in the network thread or in the refresh listener
when waiting for an active shard. This allows queries that rewrite to match_no_docs to bypass the search thread pool
entirely even if the can_match phase was skipped (pre_filter_shard_size > number of shards). Coordinating nodes
don't have the ability to create empty responses so this change also ensures that at least one shard creates a full empty
response while the other can return null ones. This is needed since creating true empty responses on shards require to create
concrete aggregators which would be too costly to build on a network thread. We should move this functionality to aggregation
builders in a follow up but that would be a much bigger change.
This change is also important for #49601 since we want to add the ability to use the result of other shards to rewrite the request
of subsequent ones. For instance if the first M shards have their top N computed, the top worst document in the global queue can be pass
to subsequent shards that can then rewrite to match_no_docs if they can guarantee that they don't have any document better than the provided
one.

jimczi committed:
add serialization test

jimczi committed:
iter

jimczi committed:
Merge branch 'master' into rewrite_shard_request_no_rejection

jimczi committed:
fix bwc issue

jimczi committed:
address review

jimczi committed:
adapt test

jimczi committed:
fix test

jimczi committed:
fix topNSize when size is reset to 0

jimczi committed:
add more comments

jimczi committed:
Merge branch 'master' into rewrite_shard_request_no_rejection

jimczi committed:
more review

jimczi committed:
initial commit

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
iter

jimczi committed:
iter

jimczi committed:
Merge branch 'rewrite_shard_request_no_rejection' into distributed_time_sort

jimczi committed:
iter2

elasticmachine commented:
Pinging @elastic/es-distributed (:Distributed/Distributed)

elasticmachine commented:
Pinging @elastic/es-search (:Search/Search)

jimczi committed:
remove unrelated change

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
fix last merge

jimczi committed:
fix rest test

jimczi committed:
another fix

mayya-sharipova reviewed (approved):
Overall, LGTM, I just left a couple of comments.

mayya-sharipova commented on rest-api-spec/src/main/resources/rest-api-spec/test/search/140_pre_filter_search_shards.yml:
```diff
@@ -180,5 +178,47 @@ setup:
+
+  - match: { _shards.total: 3 }
+  - match: { _shards.successful: 3 }
+  - match: { _shards.skipped: 0 }
```
I am wondering if we send `MatchNoneQueryBuilder()` request to a shard, or `size =0` for a request containing query part, can we say that we skipped this shard?

jimczi replied:
I have mixed feelings about this one. It would provide a valuable feedback for users since rewriting to MatchNoneQueryBuilder means that we can avoid the search threadpool but that information should be materialized for users with fast response time too. It will also make the counter non-deterministic so I wonder how much confusion this would bring.

mayya-sharipova replied:
I see, makes sense to keep it as it now.

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
address review

jimczi committed:
adapt serialization version before backport

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -72,4 +85,38 @@ protected void onShardGroupFailure(int shardIndex, Exception exc) {
         return new FetchSearchPhase(results, searchPhaseController, context);
     }
+
+    ShardSearchRequest rewriteShardRequest(ShardSearchRequest request) {
```
I am not sure I follow. Are you talking of handling queries sorted by _score ? We can probably propagate the global min competitive score up to the query collector so that wouldn't require any rewrite.

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchPhaseController.java:
```diff
@@ -683,6 +683,15 @@ private synchronized void consumeInternal(QuerySearchResult querySearchResult) {
+        public TopDocs getBufferTopDocs() {
+            if (hasTopDocs) {
+                synchronized (this) {
+                    return topDocsBuffer[0];
```
It's correct since any top hits that contains the requested size can be used but I agree that checking all top hits no matter what the buffer size could yield significant improvements. I was expecting to lower the batched_reduce_size by default in # so that this won't make a big difference but there are some discussions regarding the impact of setting this to a low value so making the change that you propose would make this decision less important for this optim. We also expect that this optimization will be really effective when the sort values are not randomly distributed so we could just check the worst hit of each shard and keep the minimum one which can be done linearly. I'll work on the change.

jimczi replied:
I pushed https://github.com/elastic/elasticsearch/pull/51852/commits/bf8ebde8430f12d9698f72e8d3ca54e61070993a

jimczi commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -72,4 +85,38 @@ protected void onShardGroupFailure(int shardIndex, Exception exc) {
+            SearchSourceBuilder source = request.source();
+            int trackTotalHits = source.trackTotalHitsUpTo() == null ? DEFAULT_TRACK_TOTAL_HITS_UP_TO :
+                source.trackTotalHitsUpTo();
+            if (topDocs.totalHits.value >= trackTotalHits) {
```
++, that will converge more quickly to the provided trackTotalHits.

jimczi replied:
I pushed https://github.com/elastic/elasticsearch/pull/51852/commits/bf8ebde8430f12d9698f72e8d3ca54e61070993a

jimczi commented on server/src/main/java/org/elasticsearch/search/internal/ShardSearchRequest.java:
```diff
@@ -343,6 +367,25 @@ public String getDescription() {
+                    && primarySort != null
+                    && primarySort.isBottomSortWithinShard(shardContext, request.getRawBottomSortValues()) == false) {
+                newSource = newSource.shallowCopy();
+                if (newSource.trackTotalHitsUpTo() == TRACK_TOTAL_HITS_DISABLED
```
Sorry I am not following. Are you suggesting that we need another boolean ?

mayya-sharipova replied:
Sorry, you are already doing that in [SearchQueryThenFetchAsyncAction](https://github.com/elastic/elasticsearch/pull/51852/files#diff-1f43ba8663f2293d15e91bb03182fcecR127)

jimczi committed:
address review

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi commented:
Thanks for the review @mayya-sharipova . I pushed another iteration to address your comments that is ready for another round of review. Can you take a look when you have time.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -72,4 +89,67 @@ protected void onShardGroupFailure(int shardIndex, Exception exc) {
+            if (topDocs.scoreDocs.length == topDocsSize) {
+                FieldDoc cand = (FieldDoc) topDocs.scoreDocs[topDocsSize-1];
+                if (bestBottom == null ||
+                        fieldComparator.compareValues(cand.fields[0], cand.fields[0]) * reverseMul > 0) {
```
should it be `fieldComparator.compareValues(bestBottom, cand.fields[0])....`

jimczi replied:
Gloups, thanks. I renamed before pushing, this should be correct now.

jimczi committed:
fix comparator

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -72,4 +89,67 @@ protected void onShardGroupFailure(int shardIndex, Exception exc) {
+        }
+        SearchSourceBuilder source = request.source();
+        int trackTotalHits = source.trackTotalHitsUpTo() == null ? DEFAULT_TRACK_TOTAL_HITS_UP_TO : source.trackTotalHitsUpTo();
+        long totalHits = -1;
```
It is not clear for me why a default value is `-1`?

jimczi replied:
It should be 0, thanks. I pushed https://github.com/elastic/elasticsearch/pull/51852/commits/8eae737035abbf59b39194d364d68cb7a33d9fae

jimczi committed:
fix topdocs size

mayya-sharipova commented:
@jimczi  Thanks for the update, I will look into more details a little bit more today.

I wanted to clarify one thing how we do a distributed search in  `AbstractSearchAsyncAction`.  As I understood, we will always send  requests to every node's 5 shards.  Once we get some initial results,  we can do  `setRawBottomSortValues` for other shards left in pendingExecutions per node. Is this the right thinking?  If yes, do you think we should consider how to avoid sending initial requests to all nodes that may not have competitive scores?

jimczi commented:
> I wanted to clarify how we do a distributed search in AbstractSearchAsyncAction. As I understood, we will always send requests to every node's 5 shards. Once we get some initial results, we can do setRawBottomSortValues for other shards left in pendingExecutions per node. Is this the right thinking? If yes, do you think we should consider how to avoid sending initial requests to all nodes that may not have competitive scores?

This is correct but I don't think we should change the execution path. The throttling is done so that we don't flood the search queues of all nodes if the request targets thousands of nodes. The initial batch that sends 5*numNodes shard request is incompressible imo since we want to allow for good parallelism when executing on a cluster with lots of nodes. This change targets requests that targets a large number of shards so I think that's ok, for the shard search we also have the sort optim in place thanks to your previous pr.

mayya-sharipova commented on server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -351,6 +352,41 @@ public SortFieldAndFormat build(QueryShardContext context) throws IOException {
+     * Returns whether all values of the given {@link QueryShardContext#getIndexReader()} are within the
+     * primary sort value provided in the <code>rawBottomSortValues</code>.
+     */
+    public boolean isBottomSortWithinShard(QueryShardContext context, Object[] rawBottomSortValues) {
```
Is there a reason why we pass have pass an array of bottom sort values, but we use only 1st element? Should we pass only the 1st element  – `Object[] rawBottomSortValue`?

mayya-sharipova commented:
@jimczi  Thanks Jim, your PR after the last commits, LGTM

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
iter

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
fix final merge when some shard responses have their size artifically rewrote to 0

jimczi committed:
check style

jimczi committed:
return topfielddocs if the shard search request rewrites the size to 0

jimczi commented:
@mayya-sharipova I was not very happy of the relation between this pr and the `batch_reduced_size`. This optimization already depends on the `pre_filter_shard_size` (to sort shards according to the primary sort field) so I wanted to avoid requiring a change in the incremental reduce as well. I pushed another iteration that, instead of relying on the partial merge, performs a light merge of the bottom top docs every time a shard returns a response. This should be very lightweight since we only need to merge the bottom docs. We could find a  better bottom top doc if we applied the merge to the entire top hits but I feel like this solution is enough since we're targeting time-based indices where shard values for the timestamp field are disjoint among indices. Can you take another look ?

mayya-sharipova commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -85,71 +88,59 @@ protected void onShardGroupFailure(int shardIndex, Exception exc) {
+            bottomTopDocs = toMerge;
+        } else {
+            final Sort sort = new Sort(bottomTopDocs.fields);
+            bottomTopDocs = TopFieldDocs.merge(sort, 0, 1, new TopFieldDocs[]{bottomTopDocs, toMerge}, false);
```
Previously using `fieldComparator.compareValues` looks to be faster than using `TopFieldDocs.merge` (as TopFieldDocs.merge doing a lot of stuff that we don't really need),  especially considering that `mergeTopDocs` is a synchronized function, and we want to be quite fast.

mayya-sharipova commented:
@jimczi Thanks Jim for new additions, they look good to me.  I also like the latest changes in `SearchQueryThenFetchAsyncAction` better, I just left a small comment about performance.

jimczi commented:
> I just left a small comment about performance.

The previous version was sorting on the primary field only. `TopFieldDocs#merge` will respect the additional sort so I opted for simplicity here. I agree that it should be fast but we're doing a merge of the bottom document for each shard. That's different than doing a full merge that retains the entire top docs for each shard. To confirm my assumption I wrote a small micro-benchmark [here](https://github.com/elastic/elasticsearch/compare/master...jimczi:search_reduce_benchmark?expand=1). It calls `TopFieldDocs#merge` on buffers of different size, 2 is equivalent to what we do here: 
````
Benchmark                             (bufferSize)  (numShards)  (topNSize)  Mode  Cnt   Score   Error  Units
TopDocsReduceBenchmark.reduceTopHits             2       100000           1  avgt   10  21,064 ± 0,501  ms/op
TopDocsReduceBenchmark.reduceTopHits            32       100000           1  avgt   10  10,362 ± 0,928  ms/op
TopDocsReduceBenchmark.reduceTopHits           512       100000           1  avgt   10   9,010 ± 0,597  ms/op
````
In this experiment, merging `100,000` shards one by one (bufferSize=2) takes 21ms in average. Considering that this operation is done while shards requests are running I don't think that it will add any overhead to have this extra computation.

jimczi committed:
Merge branch 'master' into distributed_time_sort

jpountz commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -49,27 +66,85 @@
+                mergeTopDocs((TopFieldDocs) result.queryResult().topDocs().topDocs, result.getShardIndex());
+            }
+        } catch (Exception exc) {
+            getLogger().error("Failed to merge response", exc);
```
I don't think we should ignore errors here, what if this left bottomTopDocs in an inconsistent state?

jimczi replied:
Agreed, this was a leftover from a debug session ;).

jpountz commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -49,27 +66,85 @@
+        if (trackTotalHitsUpTo != SearchContext.TRACK_TOTAL_HITS_ACCURATE
+                && current.totalHits.value >= trackTotalHitsUpTo
+                && current.totalHits.relation == GREATER_THAN_OR_EQUAL_TO) {
+            assert request.source() != null : "source should contain a primary sort field";
```
is this assertion right? I'm not seeing where we make sure to only call this method on sorted requests.

jpountz commented on server/src/main/java/org/elasticsearch/action/search/SearchQueryThenFetchAsyncAction.java:
```diff
@@ -49,27 +66,85 @@
+    }
+
+    ShardSearchRequest rewriteShardSearchRequest(ShardSearchRequest request) {
+        final TopFieldDocs current = bottomTopDocs;
```
maybe rename to `currentBottomTopDocs` to make the below logic easier to read?

jpountz commented on server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -351,6 +352,41 @@ public SortFieldAndFormat build(QueryShardContext context) throws IOException {
+        Object maxValue = order() == SortOrder.DESC ? null : rawBottomSortValues[0];
+        try {
+            MappedFieldType.Relation relation = fieldType.isFieldWithinQuery(context.getIndexReader(), minValue, maxValue,
+                true, true, null, DEFAULT_DATE_TIME_FORMATTER.toDateMathParser(), context);
```
`rawBottomSortValues` stores the implementation-specific values while `isFieldWithinQuery` expects high-level representations of the values. For instance, I believe that this implementation as-is is incorrect for `date_nanos`, which would return a number of nanos since Epoch in `rawBottomSortValues` and then parse a number of millis in `isFieldWithinQuery`. So we should call `DocValueFormat#format` on the min/max values before passing them to `isFieldWithinQuery`.

jimczi replied:
Good catch, I pushed a change to handle this correctly. I had to pass the original docvalues format in the shard search request in order to use the correct date math parser.

jpountz commented on server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -351,6 +352,41 @@ public SortFieldAndFormat build(QueryShardContext context) throws IOException {
+                true, true, null, DEFAULT_DATE_TIME_FORMATTER.toDateMathParser(), context);
+            return relation == MappedFieldType.Relation.DISJOINT;
+        } catch (Exception exc) {
+            return false;
```
I don't like silently ignoring exceptions here, can we narrow down the exception type or remove the catch block entirely? I worry that this might hide bugs.

jimczi replied:
++, this should not be needed with the new logic that send the original date math parser. I removed it.

jpountz commented on server/src/main/java/org/elasticsearch/common/lucene/Lucene.java:
```diff
@@ -111,6 +111,7 @@
     public static final NamedAnalyzer KEYWORD_ANALYZER = new NamedAnalyzer("_keyword", AnalyzerScope.GLOBAL, new KeywordAnalyzer());
 
     public static final ScoreDoc[] EMPTY_SCORE_DOCS = new ScoreDoc[0];
+    public static final ScoreDoc[] EMPTY_FIELD_DOCS = new ScoreDoc[0];
```
why do we need both?

jimczi replied:
we don't, I removed it.

jpountz reviewed (commented):
I left some comments but the overall logic looks good.

jimczi committed:
Replace raw sort values with formatted ones to handle date and date_nanos correctly

jimczi commented:
Thanks for the review @jpountz . I pushed a change to address your comments, we now pass the formatted sort values when parsing the bottom sort values on the shards. Can you take another look ?

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
fix unset of track total hits in distributed search

jimczi committed:
fix compilation test after merging with master

jimczi committed:
Merge branch 'master' into distributed_time_sort

jpountz commented on server/src/main/java/org/elasticsearch/search/SearchSortValuesAndFormats.java:
```diff
@@ -0,0 +1,102 @@
+            } else if (sortValue instanceof Double) {
+                this.formattedSortValues[i] = sortValueFormats[i].format((double) sortValue);
+            } else {
+                this.formattedSortValues[i] = sortValue;
```
```suggestion
                assert sortValue == null : "Sort values must be a BytesRef, Long or Double, but got " + sortValue.getClass() + ": " + sortValue;
                this.formattedSortValues[i] = sortValue;
```

jpountz commented on server/src/main/java/org/elasticsearch/search/sort/FieldSortBuilder.java:
```diff
@@ -351,6 +353,49 @@ public SortFieldAndFormat build(QueryShardContext context) throws IOException {
+        if (docValueFormat instanceof DocValueFormat.DateTime) {
+            if (fieldType instanceof DateFieldType && ((DateFieldType) fieldType).resolution() == NANOSECONDS) {
+                // no matter what
+                docValueFormat = DocValueFormat.withNanosecondResolution(docValueFormat);
```
is my understanding correct that we need this for the case when a shard is mapped as a `date_nanos` field while other fields are mapped as `date` fields?

jimczi replied:
correct, I fixed the comment ;)

jpountz reviewed (commented):
The change looks good to me overall. One thing that I'm missing is better confidence that it doesn't make ES break in unexpected ways in corner cases, e.g.
 - one index mapped as `date` and another one as `date_nanos` (supported, do we do the right thing?)
 - one index mapped as `keyword` and another one as `date` (unsupported, hopefully we don't raise an even more confusing error message due to this optimization than before with this change?)
 - missing values in the top hits: are we dealing correctly with `null` sort values?

jimczi committed:
fix partial comment

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
iter

jimczi commented:
Thanks for the review @jpountz 

> One thing that I'm missing is better confidence that it doesn't make ES break in unexpected ways in corner cases, e.g.

Good call, I added specific tests for the mixed types cases that are supported (`date` mixed with `date_nanos`) . I also added tests around `null` values so I think we're covered there.

> one index mapped as keyword and another one as date (unsupported, hopefully we don't raise an even more confusing error message due to this optimization than before with this change?)

I  added protection against this case since this is not supported. In the pr we now ignore tophits if they are not compatible with the sort of the previous shards. However this also uncovered a bad bug with partial reduce that fails with exceptions. This is not in the scope of this pr since the bug pre-existed it so I opened another [issue](https://github.com/elastic/elasticsearch/issues/53411) that we can handle independently. The last changes I pushed should not add a bug on top of the existing one which is why I think we can decouple the new issue.

jimczi commented:
> One thing that I'm missing is better confidence that it doesn't make ES break in unexpected ways in corner cases, e.g.

This also uncovers another bug in the pre-sorting of shards in the `can_match` phase, I opened https://github.com/elastic/elasticsearch/pull/53397 to fix this.

jimczi committed:
Add stricter assertions

Co-Authored-By: Adrien Grand <jpountz@gmail.com>

jimczi committed:
checkstyle

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
sort values can be integers and floats (_doc and _score)

jimczi committed:
fix test to handle accepted raw sort format

jimczi committed:
Merge branch 'master' into distributed_time_sort

jimczi committed:
fix compil after master merge
